File: dotnet/Allmark/Block/IndentRule.cs
namespace Allmark.Block;

using Allmark.Types;

public static class IndentRule
{
	public static BlockRule Create()
	{
		return new BlockRule
		{
			Name = "indent",
			TestStart = TestStart,
			TestContinue = TestContinue,
		};
	}

	// TODO: Should this be built in and not a rule??
	private static bool TestStart(BlockParserState state, MarkdownNode parent)
	{
		if (state.I < state.Src.Length && Utils.IsSpace(state.Src[state.I]))
		{
			for (; state.I < state.Src.Length; state.I++)
			{
				var ch = Utils.GetChar(state.Src, state.I);
				if (ch == ' ')
				{
					// TODO: All the other spaces
					state.Indent += 1;
				}
				else if (ch == '\t')
				{
					// Set spaces to the next tabstop of 4 characters (e.g. for '  \t', set
					// the spaces to 4)
					state.Indent += 4 - (state.Indent % 4);
				}
				else
				{
					break;
				}
			}
		}

		return false;
	}

	private static bool TestContinue(BlockParserState state, MarkdownNode node)
	{
		return false;
	}
}

File: zig/src/block/indentRule.zig
const std = @import("std");
const BlockParserState = @import("../types/BlockParserState.zig").BlockParserState;
const BlockRule = @import("../types/BlockRule.zig").BlockRule;
const MarkdownNode = @import("../types/MarkdownNode.zig").MarkdownNode;
const isSpaceFn = @import("../utils/isSpace.zig").isSpace;

pub fn testStart(state: *BlockParserState, _parent: *MarkdownNode) bool {
    _ = _parent;
    if (state.i < state.src.len and isSpaceFn(state.src[state.i])) {
        while (state.i < state.src.len) {
            const char = state.src[state.i];
            if (char == ' ') {
                state.indent += 1;
            } else if (char == '\t') {
                state.indent += 4 - (state.indent % 4);
            } else {
                break;
            }
            state.i += 1;
        }
    }

    return false;
}

pub fn testContinue(_state: *BlockParserState, _node: *MarkdownNode) bool {
    _ = _state;
    _ = _node;
    return false;
}

pub const indentRule = BlockRule{
    .name = "indent",
    .testStart = testStart,
    .testContinue = testContinue,
};
